### PR TITLE
Argument parsing

### PR DESCRIPTION
**Pull request by miffi**

### CODE REVIEW

**Comment by wooosh** [comment]
Fix: #6

File: src/args.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      }
+      fill_flag_argument(value, opts[i]);
+      return;
+    }
+  }
+  log_err_final("unrecognized option %s", flag);
+}
+
+static int
+short_flag_parse(const char *flag, char *next_argv, struct Option *opts[],
```
Comment (garfr) [inline src/args.c]: This parser does not check to ensure the length of the args is the same, so the flag ``-vNONSENSE`` will be equivalent to ``-v``

File: src/args.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    case OPT_INT:
+      opt->out.integer = atoi(value);
+      argument_used = true;
+      break;
+  }
+  return argument_used;
+}
+
+static void
+long_flag_parse(const char *flag, struct Option *opts[], size_t opts_length) {
```
Comment (garfr) [inline src/args.c]: This also doesn't check for the difference between ``--dump-ast`` and ``--dump-ast-NONSENSE``.

File: src/args.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  }
+  log_err_final("unrecognized option %s", flag);
+  return 0; // unreachable, just to stop a warning
+}
+
+void
+parse_args(int argc, char *argv[], struct Option *opts[], size_t opts_length,
+           char **input_file) {
+  for (int i = 1; i < argc;) {
+    if (argv[i][0] == '-') {
```
Comment (garfr) [inline src/args.c:88]: This may cause problems if the string literal ``""`` or ``-`` is passed as an argument.

File: src/bcc2.c [hunk 4]
```diff
@@ -16,108 +17,126 @@
...
-           "-v : prints version\n"
-           "-ast : dumps ast to stdout\n"
-           "-ir : dumps ir to stdout\n"
-           "-regs : dumps registers to stdout\n"
-           "-platform : list platforms\n"
-           "-platform <platform> : selects the platform to compile for\n");
-    exit(EXIT_SUCCESS);
+  struct Option *opts[] = {&help,          &version,       &ast_dump_flag,
+                           &ir_dump_flag,  &reg_dump_flag, &platform_flag,
+                           &list_platforms};
```
Comment (garfr) [inline src/bcc2.c]: It may be worth it to make this array NULL terminated instead of computing its size with integer division.

**Review by garfr — CHANGES_REQUESTED** [review]
This is a great improvement over the previous argument and it should make adding more complicated features much easier. Just a handful of changes pop out to me.

**Review by garfr — APPROVED** [review]

**Pull request merged**